add data back to fleet

Andrea-Havron-NOAA opened the pull request:
<!---
Thanks for opening a PR. This commented text will **NOT** appear in the final PR. Toggle between Write and Preview to see what your PR will look like without the comments.
-->

# What is the feature?
* add data back to fleet

# How have you implemented the solution?
* add data objects to fleet
* add data error checks in information
* add get and set id functions to fleet interface and expose in rcpp_interface
* update demo and tests
* add distribution helper functions and tests
* add tests to fleet interface
 

# Does the PR impact any other area of the project, maybe another repo?
* tests and demo needed to be changed

Andrea-Havron-NOAA commented:
@msupernaw, can you check that the data is set up in fleet correctly? 

@Bai-Li-NOAA, can you review the new helper functions for distributions?

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' @param data_type A string specifying the type of data that the distribution
+#'   will be fit to. Options are listed in the function call, where the first
+#'   option listed, i.e., `"index"` is the default.
+#' @param module An identifier to a C++ fleet module that is linked to the data
+#'   of interest.
```
I am wondering if "module" should be the first input argument for this function and the next one?

Bai-Li-NOAA replied:
Maybe match the order listed in the Roxygen documentation with the order of the input argument in the function.

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#'   of interest.
+#' @param family A description of the error distribution and link function to
+#'   be used in the model.
+#' @param sd A list of length two. The first entry, `"value"`, stores the
+#'   initial values for the relevant standard deviations. The second entry,
+#'   `"estimated"` is a vector of booleans indicating whether or not
+#'   standard deviation is estimated.
```
It is not clear if estimated should be a vector if sd is a vector with a length greater than one.

Bai-Li-NOAA replied:
Maybe check if the user provide both `value` and `estimated`, and specify default values in the Roxygen documentation?

Andrea-Havron-NOAA replied:
added to documentation

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+  if (!(family$family %in% families)) {
+    stop("FIMS currently does not offer this distribution.")
+  }
+  if ((data_type == "agecomp" | data_type == "lengthcomp") &
```
All and or calls in if statements should be || or && to ensure that a single boolean is returned and not a vector.

Andrea-Havron-NOAA replied:
done

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+    obs_id_name <- "observed_agecomp_data_id"
+  }
+  if (family$family == "lognormal") {
+    new_module <- new(TMBDlnormDistribution)
```
When are we going to rename these to not have TMB in the name? Should I add this to the list in the M2 Quality issue #662?

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+      length(sd$value)
+    )
+    new_module$log_logsd$set_all_estimable(sd$estimated)
+    if (family$link == "log") {
+      expected <- "log_expected_index"
+    }
+    if (family$link == "identity") {
+      expected <- "expected_index"
+    }
```
Should it only say index if it is index data?

Andrea-Havron-NOAA replied:
"log_expected_index" and "expected_index" are internal names in fleet for the expected values calculated in population

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+    if (family$link == "log") {
+      expected <- "log_expected_index"
+    }
+    if (family$link == "identity") {
```
I forgot to change all of the `family$link` to `family[["link"]]`

Andrea-Havron-NOAA replied:
done

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+                                     family,
+                                     sd = list(value = 1, estimated = FALSE),
+                                     is_random_effect = FALSE) {
+  families <- c("lognormal", "gaussian")
+  if (family[["family"]] == "normal") {
+    stop("use family = gaussian() instead")
+  }
+  if (!(family[["family"]] %in% families)) {
+    stop("FIMS currently does not offer this distribution for processes.")
+  }
```
It would be good to create a helper function here and use it for this and the warnings in the previous function instead of having duplicated code.

Bai-Li-NOAA replied:
+1

Andrea-Havron-NOAA replied:
created helper function to validate user input

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+  if (!(family[["family"]] %in% families)) {
+    stop("FIMS currently does not offer this distribution for processes.")
+  }
+  if (family[["family"]] == "lognormal") {
```
Before this line it would be good to have a comment delineating that you are actually setting up the distribution here based on the type passed. That way it is clear where code should be added if a new distribution is allowed. Leave some breadcrumbs for the future reader and developer of the code. This same comment pertains to line 44 as well.

Andrea-Havron-NOAA replied:
added comments

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' \item{link}{character: the link name.}
+#' \item{linkfun}{function: the link.}
+#' \item{linkinv}{function: the inverse of the link function.}
+#' \item{mu.eta}{
+#'   TODO: document mu.eta
+#'   function: derivative \eqn{TODO}.
+#' }
```
Needs more documentation.

Also, this is a REAL BIG picture comment. {sdmTMB} also has similar functions in [families.R](https://github.com/pbs-assess/sdmTMB/blob/b146eb3644321f680870a65d6e11398af0429cd8/R/families.R#L62). It would be great if we could create a package to store these where both {sdmTMB}, {FIMS}, and any other package that wants to could use them without needing to require a complex package.

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' Multinomial family and link specification
+#'
+#' @param link link function association with family
+#' @return An object of class "family"
```
The full list that is returned needs to be documented.

Andrea-Havron-NOAA replied:
done

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+
+#' Multinomial family and link specification
+#'
+#' @param link link function association with family
```
Needs to be a complete sentence.

Andrea-Havron-NOAA replied:
done

msupernaw reviewed (approved):
I'm good with these changes. We'll need to give some attention in M2Q to the derived quantity calculations and only calculate what is needed given the provided data.

Bai-Li-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#'   initial values for the relevant standard deviations. The second entry,
+#'   `"estimated"` is a vector of booleans indicating whether or not
+#'   standard deviation is estimated.
+#' @export
```
Maybe provide a `\dontrun{}` example?

Andrea-Havron-NOAA replied:
added example

Bai-Li-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' @export
+new_data_distribution <- function(
+  module,
+  family,
```
It's unclear if the `family` input can accept multiple options. The user might want to provide a few family options corresponding to a vector of data_type (e.g., `c("index", "agecomp", "lengthcomp")`)? Including a `\dontrun{}` example and specifying the expected dimensions of the input arguments in the Roxygen documentation could help clarify this.

Andrea-Havron-NOAA replied:
added documentation to clarity dimensions of input

Bai-Li-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+      length(sd$value)
+    )
+    new_module$log_logsd$set_all_estimable(sd$estimated)
+    if (family[["link"]] == "log") {
```
The code occurs twice in the function (see Lines 64-69). Maybe make this into a helper function?

Andrea-Havron-NOAA replied:
created helper function

Bai-Li-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+  return(new_module)
+}
+
+#' Sets up a new distribution for a process
```
Maybe add documentation for the return value, similar to that of `new_data_distribution()`?
Maybe provide more details about valid options for `par` and how they relate to the `module`?
It would be great to add a `\dontrun{}` example.

Andrea-Havron-NOAA replied:
added requested documentation

Bai-Li-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' fam[["family"]]
+#' fam$link
+multinomial <- function(link = "logit") {
+  r <- list(family = "multinomial")
```
The names `r `and `f` could be more descriptive.

Bai-Li-NOAA commented on tests/testthat/test-distribution-formulas.R:
```diff
@@ -0,0 +1,89 @@
+  fishing_fleet_index_distribution <-
+    new_data_distribution(data_type = "cpue", module = fishing_fleet,
+                        family = lognormal(link = "log"),
+                        sd = list(value = rep(sqrt(log(em_input$cv.L$fleet1^2 + 1)), om_input$nyr),
```
This code has been repeated several places. Maybe extract it into a separate variable.

Andrea-Havron-NOAA replied:
done

Bai-Li-NOAA commented:
I have reviewed `R/distribution_formulas.R` and the R tests. The functions have good error handling. One thing to consider is refactoring some of the repeated code.

Andrea-Havron-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+) {
+  data_type <- match.arg(data_type)
+  families <- c("lognormal", "gaussian", "multinomial")
+  if (family[["family"]] == "normal") {
```
Each family has a default link function, which is the canonical link function based on the exponential family definition. If no link function is provided, the default argument is used, which is based on the structure of the family class from the stats package.

Andrea-Havron-NOAA replied:
but we should probably verify that the family argument is a class type of family.

Andrea-Havron-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+    new_module <- new(TMBDlnormDistribution)
+    new_module$log_logsd <- new(
+      ParameterVector,
+      log(sd$value),
```
added check

kellijohnson-NOAA replied:
But the check does not capture the case when they are both greater than one but of different lengths.

Andrea-Havron-NOAA replied:
The check_distribution_validity function throws an error in this case. See line 49 of this file.

Andrea-Havron-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+  families <- c("lognormal", "gaussian", "multinomial")
+  if (family[["family"]] == "normal") {
+    stop("use family = gaussian() instead")
+  }
```
updated to use cli styling, thanks for the suggestion!

Andrea-Havron-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,192 @@
+#' fam[["family"]]
+#' fam$link
+lognormal <- function(link = "log") {
+  r <- list(family = "lognormal")
```
done

Andrea-Havron-NOAA commented:
I've addressed all the comments. Thanks for all the great feedback, I think the helper functions are more readable now! @kellijohnson-NOAA and @Bai-Li-NOAA, let me know if you approve the changes or have any follow-up comments.

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,356 @@
+#' Validaity checks for new_data_distribution and new_process_distribution
+#' 
+check_distribution_validity <- function(args){
+  list2env(args, envir = environment())
```
Just an FYI this will lead to warning checks on CRAN, not that we are trying for CRAN right now, because there is no way to know what is present in "args".

Andrea-Havron-NOAA replied:
good catch! I like to try to make the code as CRAN ready as possible so we have less to fix later if we ever decide to submit.

Andrea-Havron-NOAA replied:
removed use of list2env

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,356 @@
+#' 
+set_expected_name <- function(family, data_type){
+  expected_name <- NA
+  if(data_type == "index" || data_type == "cpue"){
+    if(family[["family"]] == "lognormal" || family[["family"]] == "gaussian"){
+      if(family[["link"]] == "log"){
+        expected_name <- "log_expected_index"
+      } 
+      if(family[["link"]] == "identity"){
+        expected_name <- "expected_index"
+      } 
+    }
+  }
+  if(data_type == "agecomp" || data_type == "lengthcomp"){
+    expected_name <- "proportion_catch_numbers_at_age"
+  }
```
Only if statements are present, no else statement. So what happens if you do not fit inside an if statement?

Andrea-Havron-NOAA replied:
data_type uses match_args to set the value from the subset: `index, cpue, agecomp, lengthcomp`. A different data type will throw an error before this part of the code is run. Also, the default expected_name is NA, so technically that is the returned value if input does not fit inside the if statements.

Andrea-Havron-NOAA replied:
I just read that match_arg uses partial matching, so I added an additional validity check to make sure data_type is one of the four options available. I also realized we don't have checks on link functions. sdmTMB adds these checks to the family functions themselves. I added a check to throw an error if the expected_value is still NA while we work out where to put checks on link functions.

kellijohnson-NOAA commented:
@Andrea-Havron-NOAA did you want to rebase this to dev while you are doing the changes to `cli::*`?

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,370 @@
+      ))
+    }
+
+    if ((data_type == "agecomp" || data_type == "lengthcomp") &&
```
I think we can change this to `if (grepl("comp", data_type)) &&` to be more generic.

kellijohnson-NOAA commented on R/distribution_formulas.R:
```diff
@@ -0,0 +1,370 @@
+  }
+  if(!is.null(args$data_type)){
+    data_type <- args$data_type
+    data_type_names <- c("index", "cpue", "agecomp", "lengthcomp")
```
Why is "cpue" an option here?

Andrea-Havron-NOAA replied:
I wasn't sure what descriptior to use here... should it be `landings`, or should `index` be used to describe both fleet and survey data?

Andrea-Havron-NOAA commented:
> @Andrea-Havron-NOAA did you want to rebase this to dev while you are doing the changes to `cli::*`?

yes, I can work on this rebase.

Andrea-Havron-NOAA commented:
> > @Andrea-Havron-NOAA did you want to rebase this to dev while you are doing the changes to `cli::*`?
> 
> yes, I can work on this rebase.

I have addressed the most recent edits, sqashed all commits, and rebased with dev

kellijohnson-NOAA commented:
@Andrea-Havron-NOAA sorry to leave this PR open for so long but I have one more major question ... should we fix the distributions per @Bai-Li-NOAA's comments that they should be more similar, e.g., take an sd argument rather than log_logsd, etc. in this PR? Then, I have a minor question, mainly for @Bai-Li-NOAA should the functions that set up the distributions be named `setup_process_distribution()` and `setup_data_distribution()` instead of `new_*()` to match the functions that you are creating?

Bai-Li-NOAA commented:
@kellijohnson-NOAA, I'm going to chat with Andrea about the distributions today. Along with making the distribution arguments more similar, I also want to figure out which fields of the distribution object should be accessible to users.

For the R function names, please feel free to leave them as they are for now. I’ll handle the refactoring later in the branch I’m working on. I’m thinking about using `initialize_*()` instead of `setup_*()` if the function involves using `methods::new()` to initialize an object.

Andrea-Havron-NOAA commented:
> @Andrea-Havron-NOAA sorry to leave this PR open for so long but I have one more major question ... should we fix the distributions per @Bai-Li-NOAA's comments that they should be more similar, e.g., take an sd argument rather than log_logsd, etc. in this PR? Then, I have a minor question, mainly for @Bai-Li-NOAA should the functions that set up the distributions be named `setup_process_distribution()` and `setup_data_distribution()` instead of `new_*()` to match the functions that you are creating?

@kellijohnson-NOAA, @Bai-Li-NOAA and I think we need the changes from both this PR and her branch to move forward with standardizing distribution arguments. After this branch gets rebased with dev, she can rebase her branch with dev and I can create a new branch off of hers to work on these changes.
